pkg/store/limiter.go

```diff
@@ -82,16 +86,89 @@ func NewChunksLimiterFactory(limit uint64) ChunksLimiterFactory {
 	}
 }
 
-// NewSeriesLimiterFactory makes a new NewSeriesLimiterFactory with a static limit.
+// NewSeriesLimiterFactory makes a new SeriesLimiterFactory with a static limit.
 func NewSeriesLimiterFactory(limit uint64) SeriesLimiterFactory {
 	return func(failedCounter prometheus.Counter) SeriesLimiter {
 		return NewLimiter(limit, failedCounter)
 	}
 }
 
-// NewSeriesLimiterFactory makes a new NewSeriesLimiterFactory with a static limit.
+// NewBytesLimiterFactory makes a new BytesLimiterFactory with a static limit.
 func NewBytesLimiterFactory(limit units.Base2Bytes) BytesLimiterFactory {
 	return func(failedCounter prometheus.Counter) BytesLimiter {
 		return NewLimiter(uint64(limit), failedCounter)
 	}
 }
+
+type RateLimits struct {
+	SeriesPerRequest  uint64
+	SamplesPerRequest uint64
+}
+
+func (l *RateLimits) RegisterFlags(cmd extkingpin.FlagClause) {
+	cmd.Flag("store.grpc.series-limit", "The maximum series allowed for a single Series request. The Series call fails if this limit is exceeded. 0 means no limit.").Default("0").Uint64Var(&l.SeriesPerRequest)
+	cmd.Flag("store.grpc.samples-limit", "The maximum samples allowed for a single Series request, The Series call fails if this limit is exceeded. 0 means no limit. NOTE: For efficiency the limit is internally implemented as 'chunks limit' considering each chunk contains a maximum of 120 samples.").Default("0").Uint64Var(&l.SamplesPerRequest)
+}
+
+// rateLimitedStoreServer is a storepb.StoreServer that can apply rate limits against Series requests.
+type rateLimitedStoreServer struct {
```

Comment:
Should we add the classic assignment to interface type to ensure this type keeps implementing the `storepb.StoreServer` interface at compile time?

Reply:
Sure thing, done.